### DISCUSSION TITLE
Add installed location to PATH environment variable

### DISCUSSION THREAD

**Discussion by MattTheCuber** [tauri-apps/tauri#9991] [Q&A]
Hello,

Forgive my ignorance in this discussion on the internals of Tauri, I am very new to the framework.

We recently finished creating a Tauri application (a Python web app) that can be bundled as an installer for Windows. We want to bundle additional executables (created with PyInstaller) that provide CLI utilities for the system. Note, these are completely separate from and not used by the web app. My first thought was to include them in the `tauri.bundle.externalBin` configuration setting. This correctly packages the executables and copies them to the install directory on Windows. Obviously, these executables can not be run from a terminal without being in the directory or adding the install directory to the `PATH` environment variable.

What is your recommendation for adding the install directory to the `PATH` environment variable during installation? Is there a way to have a script/command run during the installation to accomplish this? Currently, we are only focused on Windows, but in the future we may want a cross-compatible solution. Maybe we are thinking about this the wrong way and there is a better alternative for bundling additional utilities?

Thank you!

**Comment by zC4sTr0** [comment]
Hi MattTheCuber,

First: Hello! ^^ Welcome! To add your install directory to the PATH environment variable during the installation of your Tauri app on Windows, you can use a custom script and make a few changes to your `tauri.conf.json` file. Here’s a more straightforward way to do it:

### Step-by-Step Guide

1. **Create a PowerShell Script**:
   - Make a script that will add your install directory to the PATH environment variable.

2. **Update `tauri.conf.json`**:
   - Configure your Tauri setup to run this script during the installation process.

### Here’s How You Can Do It

#### 1. Create the PowerShell Script

Create a file named `add_to_path.ps1` with the following content:

```powershell
# add_to_path.ps1

# Define the install directory (change this to your actual path)
$installDir = "C:\Path\To\Your\Installation\Directory"

# Get the current PATH environment variable
$oldPath = [System.Environment]::GetEnvironmentVariable("Path", [System.EnvironmentVariableTarget]::Machine)

# Check if the directory is already in PATH
if ($oldPath -notmatch [regex]::Escape($installDir)) {
    # Add the install directory to PATH
    $newPath = "$oldPath;$installDir"
    [System.Environment]::SetEnvironmentVariable("Path", $newPath, [System.EnvironmentVariableTarget]::Machine)
    Write-Output "Successfully added $installDir to PATH"
} else {
    Write-Output "$installDir is already in PATH"
}
```

#### 2. Modify Your `tauri.conf.json`

Update your `tauri.conf.json` to include running this script during installation:

```json
{
  "package": {
    "productName": "YourAppName",
    "version": "0.1.0"
  },
  "tauri": {
    "bundle": {
      "active": true,
      "targets": ["msi"],
      "externalBin": ["path/to/your/executable1", "path/to/your/executable2"],
      "windows": {
        "wix": {
          "language": ["en-US"],
          "custom": {
            "PreInstall": {
              "command": "powershell",
              "args": [
                "-ExecutionPolicy",
                "Bypass",
                "-File",
                "path/to/add_to_path.ps1"
              ]
            }
          }
        }
      }
    }
  }
}
```

Check if this anser is suitable to you and please do not forget to mark as answered ! 
Thank you

**Reply by MattTheCuber** [reply]
Hello @zC4sTr0!

Thank you for the response. I was asking ChatGPT earlier and got a similar answer using `PreInstall` which did not work. The documentation for the [Tauri configuration](https://tauri.app/v1/api/config/) does not list the `custom` or `PreInstall` keys.

Side note, I am looking to use NSIS instead of WIX. I am also trying to avoid having to write custom template NSIS or WIX files/fragments.

**Comment by MattTheCuber** [comment]
It appears Tauri v2 might include this feature: https://v2.tauri.app/reference/config/#installerhooks

**Comment by danielniccoli** [comment]
Under Windows, apps like Chrome or Firefox are typically registered and not put into path.

https://learn.microsoft.com/en-us/windows/win32/shell/app-registration#registering-applications

**Reply by MattTheCuber** [reply]
Sorry for the delayed response. This appears to be a much more Windows-centric way of adding an application to "path". I assume this would still have to be set in the installer .nsi file unfortunately.

**Reply by danielniccoli** [reply]
It does not add the app to path. It is a completely different way entirely.

**Reply by MattTheCuber** [reply]
Yes, I see that. The trick is customizing how and what app(s) are registered (such as bundled executables).
